File: src/main/java/io/github/oleksivio/telegram/bot/core/controller/annotations/parser/filter/telegram/InlineQueryFilterParser.java
package io.github.oleksivio.telegram.bot.core.controller.annotations.parser.filter.telegram;

import io.github.oleksivio.telegram.bot.api.annotations.filter.primitive.StringFilter;
import io.github.oleksivio.telegram.bot.api.annotations.filter.telegram.UserFilter;
import io.github.oleksivio.telegram.bot.core.controller.annotations.parser.filter.FilterParser;
import io.github.oleksivio.telegram.bot.core.controller.annotations.parser.finder.Finder;
import io.github.oleksivio.telegram.bot.core.controller.handler.check.Validator;
import io.github.oleksivio.telegram.bot.core.controller.handler.check.impl.UnionExtractValidator;
import org.springframework.stereotype.Component;
import io.github.oleksivio.telegram.bot.api.annotations.filter.telegram.InlineQueryFilter;
import io.github.oleksivio.telegram.bot.api.annotations.filter.telegram.LocationFilter;
import io.github.oleksivio.telegram.bot.api.model.objects.inline.InlineQuery;

import java.util.Optional;

@Component
public class InlineQueryFilterParser implements FilterParser<InlineQueryFilter, InlineQuery> {

    @Override
    public Validator<InlineQuery> createChecker(InlineQueryFilter annotation, Finder finder) {
        UnionExtractValidator<InlineQuery> unionExtractValidator = new UnionExtractValidator<>();

        UserFilter from = annotation.from();
        if (from.status().isActive()) {
            unionExtractValidator.add(in -> Optional.ofNullable(in.getFrom()), finder.find(from));
        }
        LocationFilter location = annotation.location();
        if (location.status().isActive()) {
            unionExtractValidator.add(in -> Optional.ofNullable(in.getLocation()), finder.find(location));
        }
        StringFilter query = annotation.query();
        if (query.status().isActive()) {
            unionExtractValidator.add(in -> Optional.ofNullable(in.getQuery()), finder.find(query));
        }
        StringFilter offset = annotation.offset();
        if (offset.status().isActive()) {
            unionExtractValidator.add(in -> Optional.ofNullable(in.getOffset()), finder.find(offset));
        }

        return unionExtractValidator;
    }

    @Override
    public Class<InlineQueryFilter> getAnnotationClass() {
        return InlineQueryFilter.class;
    }

}

File: core/src/main/kotlin/io/github/oleksivio/tl/kbot/core/core/controller/annotations/parser/filter/telegram/InlineQueryFilterParser.kt
package io.github.oleksivio.tl.kbot.core.core.controller.annotations.parser.filter.telegram

import io.github.oleksivio.tl.kbot.core.annotations.filter.composite.InlineQueryFilter
import io.github.oleksivio.tl.kbot.server.api.objects.inline.InlineQuery
import io.github.oleksivio.tl.kbot.core.core.controller.annotations.parser.filter.FilterParser
import io.github.oleksivio.tl.kbot.core.core.controller.annotations.parser.finder.Finder
import io.github.oleksivio.tl.kbot.core.core.controller.handler.check.Validator
import io.github.oleksivio.tl.kbot.core.core.controller.handler.check.impl.UnionExtractValidator
 

  
class InlineQueryFilterParser :
    FilterParser<InlineQueryFilter, InlineQuery> {

    override val parserAnnotationClass = InlineQueryFilter::class

    override fun createChecker(annotation: InlineQueryFilter, finder: Finder): Validator<InlineQuery> {
        val unionExtractValidator =
            UnionExtractValidator<InlineQuery>()

        annotation.validator
                .map { validatorName -> finder.find(validatorName, InlineQuery::class) }
                .forEach { validator -> unionExtractValidator.add({ it }, validator) }

        val from = annotation.from
        if (from.status.isActive) {
            unionExtractValidator.add({ it.from }, finder.find(from))
        }
        val location = annotation.location
        if (location.status.isActive) {
            unionExtractValidator.add({ it.location }, finder.find(location))
        }
        val query = annotation.query
        if (query.status.isActive) {
            unionExtractValidator.add({ it.query }, finder.find(query))
        }
        val offset = annotation.offset
        if (offset.status.isActive) {
            unionExtractValidator.add({ it.offset }, finder.find(offset))
        }

        return unionExtractValidator
    }

}

